contracts/TwabRewards.sol

```diff
@@ -0,0 +1,379 @@
+    */
+    constructor(address _owner) Ownable(_owner) {
+        emit Deployed(_owner);
+    }
+
+    /* ============ Modifiers ============ */
+
+    /// @dev Ensure that the caller is the creator of the currently active promotion.
+    modifier onlyPromotionCreator() {
+        require(msg.sender == _getCurrentPromotion().creator, "TwabRewards/only-promotion-creator");
+        _;
+    }
+
+    /* ============ External Functions ============ */
+
+    /// @inheritdoc ITwabRewards
+    function createPromotion(address _ticket, PromotionParameters calldata _promotionParameters)
+        external
+        override
+        returns (bool)
+    {
+        require(_isPromotionActive() == false, "TwabRewards/promotion-already-active");
+
+        _requireTicket(_ticket);
+
+        uint256 _nextPromotionId = _currentPromotionId + 1;
+        _currentPromotionId = _nextPromotionId;
+
+        address _token = _promotionParameters.token;
+        uint256 _tokensPerEpoch = _promotionParameters.tokensPerEpoch;
+        uint256 _numberOfEpochs = _promotionParameters.numberOfEpochs;
+
+        IERC20(_token).safeTransfer(address(this), _tokensPerEpoch * _numberOfEpochs);
+
+        Promotion memory _nextPromotion = Promotion(
+            uint32(_nextPromotionId),
+            false,
+            msg.sender,
+            _ticket,
+            _token,
+            _tokensPerEpoch,
+            _promotionParameters.startTimestamp,
+            _promotionParameters.epochDuration,
+            _numberOfEpochs
+        );
+
+        _promotions[_nextPromotionId] = _nextPromotion;
+
+        emit PromotionCreated(_nextPromotion);
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function cancelPromotion(address _to) external override onlyPromotionCreator returns (bool) {
+        require(_isPromotionActive() == true, "TwabRewards/no-active-promotion");
+        require(_to != address(0), "TwabRewards/recipient-not-zero-address");
+
+        Promotion memory _currentPromotion = _getCurrentPromotion();
+
+        IERC20 _token = IERC20(_currentPromotion.token);
+        uint256 _remainingRewards = _getRemainingRewards(_token);
+
+        if (_remainingRewards > 0) {
+            _token.safeTransfer(_to, _remainingRewards);
+        }
+
+        _promotions[_currentPromotion.id].cancelled = true;
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function extendPromotion(uint256 _numberOfEpochs) external override returns (bool) {
+        require(_isPromotionActive() == true, "TwabRewards/no-active-promotion");
+
+        Promotion memory _currentPromotion = _getCurrentPromotion();
+
+        IERC20(_currentPromotion.token).safeTransfer(
+            address(this),
+            _numberOfEpochs * _currentPromotion.tokensPerEpoch
+        );
+
+        return true;
+    }
+
+    /// @inheritdoc ITwabRewards
+    function getCurrentPromotion() external view override returns (Promotion memory) {
```

Comment:
✂️  (it should be multi-tenant)

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/3aea73376f168c614844899a41e878f5f7030534